service/storage_service.cc

```diff
@@ -2257,6 +2257,143 @@ future<> storage_service::removenode(sstring host_id_string) {
+            }
+            auto endpoint = *endpoint_opt;
+            auto tokens = tmptr->get_tokens(endpoint);
+            auto leaving_nodes = std::list<gms::inet_address>{endpoint};
+
+            // Step 1: Decide who needs to sync data
+            //
+            // By default, we require all nodes in the cluster to participate
+            // the removenode operation and sync data if needed. We fail the
+            // removenode operation if any of them is down or fails.
+            //
+            // If the user want the removenode opeartion to succeed even if some of the nodes
+            // are not available, the user has to explicitly pass a list of
+            // node that can be skipped for the operation.
+            //
+            // TODO: modify the restful api to pass the nodes to skip
+            std::vector<gms::inet_address> nodes;
+            for (const auto& x : tmptr->get_endpoint_to_host_id_map_for_reading()) {
+                seastar::thread::maybe_yield();
+                if (x.first != endpoint && std::find(skip_sync_on_nodes.begin(), skip_sync_on_nodes.end(), x.first) == skip_sync_on_nodes.end()) {
+                    nodes.push_back(x.first);
+                }
+            }
+            slogger.info("removenode[{}]: Started removenode operation, removing node={}, nodes to sync data={}, nodes to skip sync data={}", uuid, endpoint, nodes, skip_sync_on_nodes);
+
+            // Step 2: Prepare to sync data
+            std::unordered_set<gms::inet_address> nodes_unknown_verb;
+            std::unordered_set<gms::inet_address> nodes_down;
+            auto req = removenode_request{removenode_cmd::prepare, uuid, leaving_nodes, skip_sync_on_nodes};
+            try {
+                parallel_for_each(nodes, [&ss, &req, &nodes_unknown_verb, &nodes_down, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_removenode(netw::msg_addr(node), req).then([uuid, node] (removenode_response resp) {
+                        slogger.info("removenode[{}]: Got prepare response from node={}", uuid, node);
+                    }).handle_exception_type([&nodes_unknown_verb, node, uuid] (seastar::rpc::unknown_verb_error&) {
+                        slogger.warn("removenode[{}]: Node {} does not support removenode verb", uuid, node);
+                        nodes_unknown_verb.emplace(node);
+                    }).handle_exception_type([&nodes_down, node, uuid] (seastar::rpc::closed_error&) {
+                        slogger.warn("removenode[{}]: Node {} is down for removenode_cmd verb", uuid, node);
+                        nodes_down.emplace(node);
+                    });
+                }).get();
+                if (!nodes_unknown_verb.empty()) {
+                    auto msg = format("removenode[{}]: Nodes={} do not support removenode verb. Please upgrade your cluster and run removenode again.", uuid, nodes_unknown_verb);
+                    slogger.warn("{}", msg);
```

Comment:
Better to have a cluster feature for it.

Reply:
I like to avoid using one if it is possible. The features are crowded.

Reply:
I don't see a problem with crowding, and I'd like to have one method of determining whether to enable a feature during upgrades, not two.